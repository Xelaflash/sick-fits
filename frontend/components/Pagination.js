import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';
import Head from 'next/head';
import Link from 'next/link';
import PaginationStyles from './styles/PaginationStyles';
import DisplayError from './ErrorMessage';
import { perPage } from '../config';

const PAGINATION_QUERY = gql`
  query {
    _allProductsMeta {
      count
    }
  }
`;

export default function Pagination({ page }) {
  const { error, loading, data } = useQuery(PAGINATION_QUERY);

  if (loading) return 'Loading...';
  if (error) return <DisplayError error={error} />;

  const { count } = data._allProductsMeta;
  const pageCount = Math.ceil(count / perPage);

  return (
    <PaginationStyles>
      <Head>
        <title>
          Sick Fits - Page {page} of {pageCount}{' '}
        </title>
      </Head>
      {/* 1. Render the pagination links */}
      <Link href={`/products/${page - 1}`}>
        {/*  in next you can't put extra attributes on a Link component. You need to nest a a tag inside it */}
        <a aria-disabled={page <= 1}>← Prev</a>
      </Link>
      <p>
        Page {page} of {pageCount}{' '}
      </p>
      <p>{count} items in Total</p>
      <Link href={`/products/${page + 1}`}>
        <a aria-disabled={page >= pageCount}>Next →</a>
      </Link>
    </PaginationStyles>
  );
}

// 2. Allow for dynamic routing
// 3. Filter the products for the current page
// 4. Deal with cache invalidation
