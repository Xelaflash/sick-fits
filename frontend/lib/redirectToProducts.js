import { useRouter } from 'next/router';

export default function redirectToProducts() {
  const router = useRouter();
  router.push('/products');
}
