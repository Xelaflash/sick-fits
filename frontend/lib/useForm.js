import { useState, useEffect } from 'react';

export default function useForm(initial = {}) {
  // create a state object for our inputs
  const [inputs, setInputs] = useState(initial);

  // in order to fix the bug on edit a Product where we do not get the values populated on the form ==> useForm initial state is an empty object. We need to watch the values with useEffect.

  const initialValues = Object.values(initial).join('');
  // join is just to have a string to check either empty string or values of the product.

  useEffect(() => {
    // this function runs when the thinks we are watching (form values) change
    setInputs(initial);
  }, [initialValues]);

  function handleChange(e) {
    let { value, name, type } = e.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    // for the files (picture) upload
    if (type === 'file') {
      [value] = e.target.files;
    }
    setInputs({
      // copy the existing state
      ...inputs,
      [name]: value,
    });
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ''])
    );
    setInputs(blankState);
  }

  // return things we want ot surface from this custom hook
  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
