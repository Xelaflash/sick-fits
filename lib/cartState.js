import { PropTypes } from 'prop-types';
import { useState, useContext, createContext } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // This is our own custom provider. It will store data (state) and functionality (updaters).
  // It can be accessed from other levels via the consumer.
  const [cartOpen, setCartOpen] = useState(false);
  // console.log(cartOpen);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

CartStateProvider.propTypes = {
  children: PropTypes.any,
};

//  we need a custom hook to access the cart local state from the  CartStateProvider
function useCart() {
  //  We use a CONSUMER to access local state (ie a provider needs a consumer to access data (local state)
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };
