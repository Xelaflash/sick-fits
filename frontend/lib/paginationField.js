import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // tells apollo we will handle this
    read(existing = [], { args, cache }) {
      // console.log(existing, args, cache);
      const { skip, first } = args;

      // read the nbr of items on the page from the cache
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      // console.log(data);
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      //  check if existing items
      const items = existing.slice(skip, skip + first).filter((x) => x);

      // if we have items AND there aren't enough to fulfill our Perpage Items and we are on last page, just send it.
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        // we don"'t have any item, we must go to the network
        return false;
      }

      // if there are items, just return them from the cache
      if (items.length) {
        return items;
      }
      return false; // fallback to networtk in case neither ifs are working

      // when appollo is looking fro products first thing is asks the read function for those items
      // we can either:
      // 1. return the item 'cause they are already in cache
      // 2. return false from here that will trigger a network request.
    },
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      // this runs when the apollo client comes back from network with products.
      console.log('merging item from network');
      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; i += 1) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}
