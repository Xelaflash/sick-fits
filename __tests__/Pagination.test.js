import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import Pagination from '../components/Pagination';
import { makePaginationMocksFor } from '../lib/testUtils';

describe('<Pagination />', () => {
  it('1. Displays a loading message', () => {
    const { container } = render(
      <MockedProvider mocks={makePaginationMocksFor(1)}>
        <Pagination />
      </MockedProvider>
    );
    expect(container).toHaveTextContent('Loading...');
  });
  it('2. Renders pagination for 18 items', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await screen.findByTestId('pagination');
    // debug();
    expect(container).toHaveTextContent('Page 1 of 5');
    expect(container).toMatchSnapshot();
  });
  it('3. Disable prev page on page 1', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(12)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await screen.findByTestId('pagination');
    // we use regex
    const prevBtn = screen.getByText(/Prev/);
    const nextBtn = screen.getByText(/Next/);
    expect(prevBtn).toHaveAttribute('aria-disabled', 'true');
    expect(nextBtn).toHaveAttribute('aria-disabled', 'false');
  });
  it('4. Disable next page on last page (3)', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(16)}>
        <Pagination page={4} />
      </MockedProvider>
    );
    await screen.findByTestId('pagination');
    // we use regex
    const prevBtn = screen.getByText(/Prev/);
    const nextBtn = screen.getByText(/Next/);
    expect(prevBtn).toHaveAttribute('aria-disabled', 'false');
    expect(nextBtn).toHaveAttribute('aria-disabled', 'true');
  });
  it('5. Enables all on middle pages', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(20)}>
        <Pagination page={2} />
      </MockedProvider>
    );
    await screen.findByTestId('pagination');
    // we use regex
    const prevBtn = screen.getByText(/Prev/);
    const nextBtn = screen.getByText(/Next/);
    expect(prevBtn).toHaveAttribute('aria-disabled', 'false');
    expect(nextBtn).toHaveAttribute('aria-disabled', 'false');
  });
});
