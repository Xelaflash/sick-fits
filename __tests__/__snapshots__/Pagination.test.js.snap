// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> 2. Renders pagination for 18 items 1`] = `
<div>
  <div
    class="PaginationStyles-sc-aduuar-0 ZXaJx"
    data-testid="pagination"
  >
    <a
      aria-disabled="true"
      href="/products/0"
    >
      ← Prev
    </a>
    <p>
      Page 
      1
       of 
      5
       
    </p>
    <p>
      18
       items in Total
    </p>
    <a
      aria-disabled="false"
      href="/products/2"
    >
      Next →
    </a>
  </div>
</div>
`;
